test/common/testlib.py

```diff
@@ -1131,33 +1131,56 @@ class _DebugOutcome(unittest.case._Outcome):
-                        # strip off the two topmost frames for testPartExecutor and TestCase.run(); uninteresting and breaks naughties
-                        traceback.print_exception(exc_info[0], exc_info[1], exc_info[2].tb_next.tb_next)
-                        try:
-                            test_case.snapshot("FAIL")
-                            test_case.copy_js_log("FAIL")
-                            test_case.copy_journal("FAIL")
-                            test_case.copy_cores("FAIL")
-                        except (OSError, RuntimeError):
-                            # failures in these debug artifacts should not skip cleanup actions
-                            sys.stderr.write("Failed to generate debug artifact:\n")
-                            traceback.print_exc(file=sys.stderr)
-
-                        if opts.sit:
-                            sit(test_case.machines)
-
-        return super().testPartExecutor(test_case, isTest)
+        def failureHandler(exc_info, python_311=False):
+            if python_311:
+                # exc_info is now a string
+                print(exc_info, file=sys.stderr)
+            else:
+                # strip off the two topmost frames for testPartExecutor and TestCase.run(); uninteresting and breaks naughties
+                traceback.print_exception(exc_info[0], exc_info[1], exc_info[2].tb_next.tb_next)
+            try:
+                test_case.snapshot("FAIL")
+                test_case.copy_js_log("FAIL")
+                test_case.copy_journal("FAIL")
+                test_case.copy_cores("FAIL")
+            except (OSError, RuntimeError):
+                # failures in these debug artifacts should not skip cleanup actions
+                sys.stderr.write("Failed to generate debug artifact:\n")
+                traceback.print_exc(file=sys.stderr)
+
+            if opts.sit:
+                sit(test_case.machines)
+
+        superResult = super().testPartExecutor(test_case, isTest)
+        ran_debug = hasattr(test_case, "_ran_debug")
+
+        if ran_debug:
+            return superResult
+
+        # Python < 3.11 (Outcome class does not have a errors attribute anymore
+        # https://github.com/python/cpython/commit/664448d81f41c5fa971d8523a71b0f19e76cc136#diff-c6fe1ffe930def48a6adf1fa99b974737bac586fdacccacd1474e7b2f11370ebL79
+        if hasattr(self, 'errors'):
+            if self.errors and not isTest:
+                (err_case, exc_info) = self.errors[-1]
+                if exc_info and isinstance(test_case, MachineCase):
+                    assert err_case == test_case
+                    setattr(test_case, "_ran_debug", True)
+                    failureHandler(exc_info, False)
+        elif self.result.errors or self.result.failures:
+            errors = []
+            if self.result.errors:
+                errors.append(self.result.errors[-1])
+            if self.result.failures:
+                errors.append(self.result.failures[-1])
```

Comment:
Or maybe a much more elegant way to do this would be to itertools.chain() .errors and .failures, and use a list comprehension to filter out all items that have `test_case == err_case`. The result should contain at most one item (let's assert that), and that's the one to process then. This should also work with the old API.

WDYT?

Reply:
I've opted for itertools.chain() not sure if they can be combined, that requires some follow up. Also moved the isinstance check out as it can be applied earlier and for both.